### ISSUE TITLE
Log validation loss while also evaluating.

### ISSUE THREAD

**Issue by lorinczszabolcs** [open-mmlab/mmsegmentation#1711]
Hi!

My question would be how to log validation loss while also being able to evaluate on the data set.

The error I am running into is that if I add ('val', 1) to the workflow, it won't have the labels loaded, since originally the test_pipeline is used for validation as well. On the other hand, when I use a separate val_pipeline, loading labels as well, then the validation loss can't be calculated because the multiscaleflip augmentation returns a list instead of a tensor, and if I remove the multiscale flip augmentation from the validation pipeline, the final evaluation will not work, since that expects an input of type list.

I know that there were several related issues, but all of them solved separate subproblems in my opinion, and the underlying main problem is that different types are expected as input during evaluation and validation loss logging. Let me know, if I can make things more clear.

Thank you in advance!

**Comment by lorinczszabolcs** [comment]
Solved by:

1. Adding ```workflow = [('train', 1), ('val', 1)]```
2. Separate ```val_pipeline``` that doesn't contain ```MultiScaleFlipAug``` and contains ```LoadAnnotations``` and collecting gt segmentations.
3. In custom data set, ```prepare_test_scan``` was changed to also load annotation info, similarly to ```prepare_train_scan``` and after calling ```self.pipeline(results)```, the elements of the dictionary need to be wrapped in a list before returning them.
4. In training script, added ```datasets = [build_dataset(cfg.data.train), build_dataset(cfg.data.val)]```

**Comment by javadmozaffari** [comment]
Hi,
for number 3, where can I find 'prepare_test_scan' and 'self.pipeline(results)'

**Comment by javadmozaffari** [comment]
HI,
How can I wrap the elements of the dictionary in a list?

**Comment by stone-cloud** [comment]
@lorinczszabolcs 
I met a similar problem. I tried your solution, but I have doubts about the third step, can you release the relevant modified code?Thank you very much！
